File: remove-all-adjacent-duplicates-in-string-ii/remove-all-adjacent-duplicates-in-string-ii.py
class Solution:
    def removeDuplicates(self, s: str, k: int) -> str:
        stack = []
        
        for c in s:
            if not stack or stack[-1][0] != c:
                stack.append([c,1])
            elif stack[-1][1] == k - 1:
                stack.pop()
            else:
                stack[-1][1] += 1
        
        return "".join([x[0] * x[1] for x in stack])
                

File: remove-all-adjacent-duplicates-in-string-ii/remove-all-adjacent-duplicates-in-string-ii.java
class Solution {
    public String removeDuplicates(String s, int k) {
        Deque<Character> stack = new ArrayDeque<>();
        Deque<Integer> count = new ArrayDeque<>();
        
        for (char c : s.toCharArray()) {
            if (!stack.isEmpty() && c == stack.getLast()) {
                count.addLast(count.pollLast() + 1);
                
                
                if (count.peekLast() == k) {
                    count.pollLast();
                    stack.pollLast();
                }
            } else {
                stack.addLast(c);
                count.addLast(1);
            }
        }
        
        StringBuilder sb = new StringBuilder();
        while(!stack.isEmpty()) {
            char c = stack.pollFirst();
            int num = count.pollFirst();
            for (int i = 0; i < num; i++) {
                sb.append(c);
            }
        }
        
        return sb.toString();
    }
}